File: tools/encoders/rle_encoder/on_text/rleEncoder.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <pthread.h>
#include <unistd.h>
#include <vector>
#include <chrono>

struct threadData {
    std::string inputStr;
    std::string rleStr;
    std::vector<int> rleFreq;
    std::chrono::time_point<std::chrono::high_resolution_clock> startTime, endTime;
};

void* rle_encode(void* ptr) {
    threadData* rleData = (struct threadData*) ptr;  // Cast the void pointer to threadData*

    rleData->startTime = std::chrono::high_resolution_clock::now();
    std::string str = rleData->inputStr;
    std::string rle;
    std::vector<int> freq;
    int count;

    for (size_t i = 0; i < str.length(); ++i) {
        count = 1;
        while (i + 1 < str.length() && str[i] == str[i + 1]) {
            ++count;
            ++i;
        }
        if (count > 1) {
            rle += str[i];
            rle += str[i];
            freq.push_back(count);

            rle += "[";
            rle += std::to_string(count);
            rle += "]";
        } else {
            rle += str[i];
        }
    }

    rleData->rleStr = rle;
    rleData->rleFreq = freq;

    rleData->endTime = std::chrono::high_resolution_clock::now();
    return nullptr;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <input_file>" << std::endl;
        return 1;
    }

    std::ifstream inputFile(argv[1]);
    if (!inputFile.is_open()) {
        std::cerr << "Error: Could not open file " << argv[1] << std::endl;
        return 1;
    }

    std::vector<std::string> strVect;
    std::string line;

    while (std::getline(inputFile, line)) {
        if (!line.empty()) {
            strVect.push_back(line);
        }
    }
    inputFile.close();

    size_t numStrings = strVect.size();
    threadData* data = new threadData[numStrings];
    pthread_t* threads = new pthread_t[numStrings];

    for (size_t i = 0; i < numStrings; ++i) {
        data[i].inputStr = strVect[i];
        if (pthread_create(&threads[i], nullptr, rle_encode, &data[i])) {
            std::cerr << "Error: Could not create thread" << std::endl;
            delete[] data;
            delete[] threads;
            return 1;
        }
    }

    for (size_t i = 0; i < numStrings; ++i) {
        pthread_join(threads[i], nullptr);
    }

    for (size_t i = 0; i < numStrings; ++i) {
        std::cout << "Input string: " << data[i].inputStr << std::endl;
        std::cout << "RLE String: " << data[i].rleStr << std::endl;
        std::cout << "RLE Frequencies: ";
        for (int freq : data[i].rleFreq) {
            std::cout << freq << " ";
        }
        std::cout << std::endl;

        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
                            data[i].endTime - data[i].startTime).count();
        std::cout << "Execution Time: " << duration << " ms" << std::endl;
        std::cout << "Memory Saved(in bytes): " << data[i].inputStr.size() - data[i].rleStr.size() << std::endl;
        std::cout << std::endl;
    }

    delete[] data;
    delete[] threads;
    return 0;
}

File: tools/encoders/rle_encoder/on_text/rle_encoder.py
import os
import sys
import threading
from time import time


class ThreadData:
    def __init__(self, input_str):
        self.input_str = input_str
        self.rle_str = ""
        self.rle_freq = []
        self.time_tuple = None
        self.mem_saved = 0


def rle_encode(thread_data):
    start_time = time()
    print(f"Thread for '{thread_data.input_str}' started.")

    input_str = thread_data.input_str
    rle_str = ""
    rle_freq = []

    i = 0
    while i < len(input_str):
        count = 1
        while i + 1 < len(input_str) and input_str[i] == input_str[i + 1]:
            count += 1
            i += 1

        if count > 1:
            rle_str += input_str[i] * 2
            rle_freq.append(count)
            rle_str += "["
            rle_str += str(count)
            rle_str += "]"
        else:
            rle_str += input_str[i]
        i += 1

    thread_data.rle_str = rle_str
    thread_data.rle_freq = rle_freq

    thread_data.mem_saved = len(thread_data.input_str) - len(thread_data.rle_str)
    end_time = time()
    thread_data.time_tuple = (start_time, end_time)
    print(f"Thread for '{thread_data.input_str}' finished.")


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 your_script.py input/test.txt")
        sys.exit(1)

    # go to input folder
    input_file = sys.argv[1]
    if not os.path.isfile(input_file):
        print(f"Error: File '{input_file}' not found.")
        sys.exit(1)

    # read input strings from input dir/folder
    str_list = []
    with open(input_file, "r") as file:
        for line in file:
            line = line.strip()
            if line:
                str_list.append(line)

    # create thread data objects and threads
    thread_data_list = [ThreadData(input_str) for input_str in str_list]
    threads = []

    # create and start threads
    for data in thread_data_list:
        thread = threading.Thread(target=rle_encode, args=(data,))
        threads.append(thread)
        thread.start()

    # wait for all threads to finish
    for thread in threads:
        thread.join()

    # print metrics
    for data in thread_data_list:
        print(f"Input string: {data.input_str}")
        print(f"RLE String: {data.rle_str}")
        print(f"RLE Frequencies: {' '.join(map(str, data.rle_freq))}")
        print(f"Time (start, finish): {data.time_tuple}")
        print(f"Execution Time: it took: {data.time_tuple[1] - data.time_tuple[0]}s")
        print(f"Memory Saved (in bytes): {data.mem_saved}\n")


if __name__ == "__main__":
    main()
